src/signature.sh

```diff
@@ -0,0 +1,402 @@
+{
+  local sha="$1"
+  local formatted_message
+
+  # Check if given value is a valid commit reference
+  if [[ $(git cat-file -t "$sha" 2> /dev/null) != 'commit' ]]; then
+    return 22 # EINVAL
+  fi
+
+  # The 'Fixes:' trailer line must follow a format defined by Linux Kernel developers.
+  # Fixes: e21d2170f366 ("video: remove unnecessary platform_set_drvdata()")
+  formatted_message=$(git log -1 "$sha" --oneline --abbrev-commit --abbrev=12 \
+    --format="%h (\\\"%s\\\")")
+
+  printf '%s' "$formatted_message"
+}
+
+# It gets user.name and user.email from git's configuration.
+# If either the name or email are not configured then this
+# function will return an error code. Otherwise it will output
+# the formatted name and email.
+#
+# Return:
+# In case of successful return 0 and prints a properly formatted
+# output, otherwise, return 1.
+function format_name_email_from_user()
+{
+  local user_name
+  local user_email
+
+  user_name="$(git config user.name)"
+  user_email="$(git config user.email)"
+
+  # If user doesn't have either a name or email configured with
+  # git then they must provide an argument
+  if [[ -z "$user_name" || -z "$user_email" ]]; then
+    return 1
+  fi
+
+  printf '%s' "${user_name} <${user_email}>"
+}
+
+# This function writes all the trailer lines stored in @all_trailers
+# into either a patch file or commit. It prints a warning if the
+# given argument is neither a valid patch path nor a commit reference
+# and then ignores it.
+#
+# @patch_or_sha Holds either a patch path or commit SHA
+# @flag Used to specify how 'cmd_manager' will be executed
+function write_all_trailers()
+{
+  local patch_or_sha="$1"
+  local flag="$2"
+  local cmd
+
+  for trailer in "${all_trailers[@]}"; do
+    # Check if given argument is either a patch or valid commit reference,
+    # then build the correct command.
+    if [[ "$(git cat-file -t "$patch_or_sha" 2> /dev/null)" == 'commit' ]]; then
+      cmd="git commit --quiet --amend --no-edit --trailer \"${trailer}\""
+      # Only call 'git rebase' if user is trying to write multiple commits
+      if [[ "$(git rev-parse "$patch_or_sha")" != "$(git rev-parse HEAD)" ]]; then
+        cmd="git rebase ${patch_or_sha} --exec '${cmd}' 2> /dev/null"
+      fi
+    elif is_a_patch "$patch_or_sha"; then
+      cmd="git interpret-trailers ${patch_or_sha} --in-place --trailer \"${trailer}\""
+    else
+      warning "Unmatched patch or commit. Ignoring: ${patch_or_sha}"
+      continue
+    fi
+
+    cmd_manager "$flag" "$cmd"
+  done
+}
+
+# This function gets raw data and based on that fill out the options values to
+# be used in another function.
+#
+# Return:
+# In case of successful return 0, otherwise, return 22.
+function parse_signature_options()
+{
+  local long_options='add-signed-off-by::,add-reviewed-by::,add-acked-by::,add-fixes::'
+  long_options+=',add-tested-by::,add-reported-by::,add-co-developed-by::,verbose'
+  local short_options='s::,r::,a::,f::,t::,R::,C::'
```

Comment:
That actually happened to me multiple times when I was testing it myself. I had to use `git reset --hard <SHA>` to backup every time that happened. I've been thinking of how to properly handle this, but didn't have too much free time recently to work on it. But I'm going to try come up with ideas to improve robustness and avoid this really annoying issue.

Reply:
The solution I'll be using for now is giving an error to the user and stop the feature from running any further if an invalid commit or patch is detected within the arguments. Not sure if it's the best solution, but at least it will help us to avoid this catastrophic scenario where many undesirable changes are made and the user might not have a proper way to back it all up.